python/main.py

```diff
@@ -0,0 +1,583 @@
+
+        # handle type_comment
+
+        return fd
+
+    def visit_Lambda(self, node):
+        debug_print(ast.dump(node))
+        raise NotImplementedError
+
+    def visit_arguments(self, node):
+        debug_print(ast.dump(node))
+        plist = []
+        for p in node.posonlyargs:
+            raise NotImplementedError
+        for p in node.args:
+            plist.append(self.visit(p))
+        if node.vararg is not None:
+            raise NotImplementedError
+        for p in node.kwonlyargs:
+            raise NotImplementedError
+        for p in node.kw_defaults:
+            raise NotImplementedError
+        if node.kwarg is not None:
+            raise NotImplementedError
+        for p in node.defaults:
+            pass
+            # raise NotImplementedError
+            # no cpg support???
+        return plist
+
+    def visit_arg(self, node):
+        debug_print(ast.dump(node))
+        pvd = ParamVariableDeclaration()
+        self.add_loc_info(node, pvd)
+        pvd.setName(node.arg)
+        # TODO handle annotation
+        return pvd
+
+    def visit_Return(self, node):
+        debug_print(ast.dump(node))
+        r = ReturnStatement()
+        self.add_loc_info(node, r)
+        r.setReturnValue(self.visit(node.value))
+        return r
+
+    def visit_Yield(self, node):
+        debug_print(ast.dump(node))
+        raise NotImplementedError
+    def visit_YieldFrom(self, node):
+        debug_print(ast.dump(node))
+        raise NotImplementedError
+    def visit_Global(self, node):
+        debug_print(ast.dump(node))
+        if len(node.names) != 1:
+            raise NotImplementedError
+        ref = DeclaredReferenceExpression()
+        self.add_loc_info(node, ref)
+        ref.setName(node.names[0])
+        return ref
+
+    def visit_Nonlocal(self, node):
+        debug_print(ast.dump(node))
+        raise NotImplementedError
+
+    def visit_ClassDef(self, node):
+        debug_print(ast.dump(node))
+        rec = RecordDeclaration()
+        self.add_loc_info(node, rec)
+        # name
+        rec.setName(node.name)
+        # bases
+        if len(node.bases) == 0:
+            pass
+        elif len(node.bases) == 1:
+            b = node.bases[0]
+            if isinstance(b, ast.Attribute):
+                tname = "%s.%s" % (b.value.id, b.attr)
+                t = TypeParser.createFrom(tname, True)
+            else:
+                raise NotImplementedError
+        else:
+            raise NotImplementedError
+        rec.setSuperClasses([t])
+        if len(node.keywords) != 0:
+            raise NotImplementedError
+        #if node.starargs is not None:
+        #    raise NotImplementedError
+        #if node.kwargs is not None:
+        #    raise NotImplementedError
+        # body
+        for b in node.body:
+            if isinstance(b, ast.FunctionDef):
+                fd = self.visit_FunctionDef(b, returnMethod=True)
+                rec.addMethod(fd)
```

Comment:
rather than adding declarations directly, the following flow should be used:

The the scope of the class / record declaration by using ScopeManager::enterScope(rec). Then afterwards within the scope of the record, use ScopeManager::addDeclaration(fd). This applies to all *declarations*. The scope manager then handles the rest automagically, by dispatching it back to the class associated with the scope, i.e. a RecordDeclaration, which then decides how to add it. This is not really documented, maybe we need a "How to write a language frontend" guide somewhere @konradweiss

Reply:
Yes, I talked with @konradweiss and I'm using the ScopeManager now.